### ISSUE TITLE
Overcome (cosmetic) shutdown error on live-ISOs / Untangling of overlayfs root mount on shutdown

### ISSUE THREAD

**Issue by eMPee584** [systemd/systemd#17988]
So, after spending several hours (and then some) (and then some) trying to axe that line
`[FAILED] Failed unmounting /run/live/medium.` from the shutdown log of our debian-based [live-distro](https://github.com/fsfw-dresden/usb-live-linux), at least I now _kind of_ understand the problem, but have not been successful in solving it yet.
First I tried adding `override.conf` files for several of the mounts, making them depend on the first of them, as in
```
[Unit]
RequiresMountsFor=/run/live/medium
```
like was recommended in another issue regarding a seemingly similar issue about unmounting overlay FSs on shutdown.
Until finally recognizing, that the actual issue here is a bit more knotty.. While the details listed here are specific to Debian (& derivatives), I've seen the same issue f.e. with a Manjaro live ISO..

So the package `live-boot` installs scripts into the `initramfs` that run a `setup_unionfs()` function which mounts an overlayfs mount on `${rootmnt}` with the upperdir on `/run/live/overlay`. After `/run` is move-mounted to `${rootmnt}`, this folder becomes the FS root when at the end of these scripts `run-init` (from `klibc-utils`) is invoked at https://salsa.debian.org/kernel-team/initramfs-tools/-/blob/buster/init#L328
`exec run-init ${drop_caps} "${rootmnt}" "${init}" "$@" <"${rootmnt}/dev/console" >"${rootmnt}/dev/console" 2>&1`
Then, the init system, systemd in this case, takes over.
This will result in a mount hierarchy as such:
```
TARGET                                             SOURCE              FSTYPE          OPTIONS
/                                                  overlay             overlay         rw,noatime,lowerdir=/run/live/rootfs/filesystem.squashfs/,
│                                                                                      upperdir=/run/live/overlay/rw,workdir=/run/live/overlay/work
├─/run                                             tmpfs               tmpfs           rw,nosuid,noexec,relatime,size=306516k,mode=755
│ ├─/run/live/medium                               /dev/sr0            iso9660         ro,noatime,nojoliet,check=s,map=n,blocksize=2048
│ ├─/run/live/rootfs/filesystem.squashfs           /dev/loop0          squashfs        ro,noatime
│ ├─/run/live/overlay                              tmpfs               tmpfs           rw,noatime,mode=755
```
where the root FS mount actually depends on a lower path. Thus, on shutdown:
```
Failed to unmount /run/live/medium: Device or resource busy
systemd-shutdown[1]: Could not detach loopback /dev/loop0: Device or resource busy
systemd-shutdown[1]: Failed to finalize file systems, loop devices, ignoring.
```
I tried to create an `untangle-root-overlay.service`:
```
[Unit]
Description=UNTANGLE root overlay
DefaultDependencies=no
Requires=-.mount run-live-rootfs-filesystem.squashfs.mount run-live-overlay.mount
After=-.mount run-live-rootfs-filesystem.squashfs.mount run-live-overlay.mount
Conflicts=shutdown.target halt.target

[Service]
Type=oneshot
RemainAfterExit=yes
ExecStart=/bin/true
ExecStop=mount -v --make-rprivate /
ExecStop=mkdir -pv /run/live/rootfs/overlay
ExecStop=pivot_root /run/live/rootfs/filesystem.squashfs /run/live/rootfs/overlay
```
but even though, after lots of poking (and having run `systemctl daemon-reload`, `systemctl start untangle-root-overlay.service`), I could not make the `ExecStop` commands be run.. So advice on making it trigger (or actually work) is welcome, as well as any general approach to make that unnerving `[FAILED]` on live-system shutdown disappear..

Addendum: the current Fedora live disc seems to work a bit different; no overlayfs on / and no failure message on shutdown.. and some interesting lines:
```
[  OK  ] Stopped Restore /run/initramfs on shutdown.                                                                                               
[…]
systemd-shutdown[1]: Successfully changed into root pivot.
```
though it also emits a debug message about not being able to detach loopback devices (harmless).

**Comment by poettering** [comment]
Hmm, it's not clear to me how we should ever be able to detect these relationships automatically. I don't think we have enough information exposed in /proc/self/mountinfo to figure this out.

What I don't get: why does debian set things up this way? It looks like a way to intentionally make automatic disassembly hard. having a submount that cannot be unmounted unless a mount *above* it is unmounted first is just, well, mean ;-)

i.e. isn't the simpler and better fix to change debian to not set things up that way? Specifically, why expose the "components" of the overlayfs at all? why not just leave them outside the tree when pivotting into the overlayfs?

**Comment by stevenshiau** [comment]
It seems systemd 248 has fixed this issue?
I tired systemd (248.3-1ubuntu1) from Ubuntu impish and systemd (248.3-1) from Debian experimental, this is no such error messages anymore:
Failed to unmount /run/live/medium: Device or resource busy
         systemd-shutdown[1]: Could not detach loopback /dev/loop0: Device or resource busy
         systemd-shutdown[1]: Failed to finalize file systems, loop devices, ignoring.

Not sure if the error messages are hidden or is this issue really fixed...

**Comment by poettering** [comment]
Fixed by 6dc68a00cfc816678fd713b12ae2a4cf2ae6da85. Closing

**Comment by eworm-de** [comment]
This should be a generic solution, no? I built an image with `254-rc2` with Arch's [archiso](https://gitlab.archlinux.org/archlinux/archiso). Looks like it is still suffering this.
